src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/core/AbstractCraftingRequestResolver.java

```diff
@@ -0,0 +1,140 @@
+    {
+        return building instanceof AbstractBuildingWorker && canBuildingCraftStack((AbstractBuildingWorker) building, request.getRequest().getStack());
+    }
+
+    public abstract boolean canBuildingCraftStack(@NotNull final AbstractBuildingWorker building, ItemStack stack);
+
+    @Nullable
+    @Override
+    public List<IToken<?>> attemptResolveForBuilding(
+      @NotNull final IRequestManager manager, @NotNull final IRequest<? extends Stack> request, @NotNull final AbstractBuilding building)
+    {
+        final AbstractBuildingWorker buildingWorker = (AbstractBuildingWorker) building;
+        return attemptResolveForBuildingAndStack(manager, buildingWorker, request.getRequest().getStack());
+    }
+
+    @Nullable
+    protected List<IToken<?>> attemptResolveForBuildingAndStack(@NotNull final IRequestManager manager, @NotNull final AbstractBuildingWorker building, final ItemStack stack)
+    {
+        final IRecipeStorage fullfillableCrafting = building.getFirstFullFillableRecipe(stack);
+        if (fullfillableCrafting != null)
+        {
+            return ImmutableList.of();
+        }
+
+        final IRecipeStorage craftableCrafting = building.getFirstRecipe(stack);
+        if (craftableCrafting == null)
+        {
+            return null;
+        }
+
+        return createRequestsForRecipe(manager, building, stack, craftableCrafting);
+    }
+
+    protected int calculateMaxCraftingCount(@NotNull final ItemStack outputStack, @NotNull final IRecipeStorage storage)
+    {
+        //Calculate the initial crafting count from the request and the storage output.
+        int craftingCount =
+          (int) Math.ceil(Math.max(ItemStackUtils.getSize(outputStack), ItemStackUtils.getSize(storage.getPrimaryOutput())) / ItemStackUtils.getSize(storage.getPrimaryOutput()));
+
+        //Now check if we excede an ingredients max stack size.
+        for(final ItemStack ingredient : storage.getInput())
+        {
+            //Calculate the input count for the ingredient.
+            final int ingredientInputCount = ItemStackUtils.getSize(ingredient) * craftingCount;
+            //Check if we are above the max stacksize.
+            if (ingredientInputCount > ingredient.getMaxStackSize())
+            {
+                //Recalculate the crafting limit using the maxstacksize of the ingredient.
+                craftingCount =
+                  (int) Math.ceil(Math.max(ingredient.getMaxStackSize(), ItemStackUtils.getSize(storage.getPrimaryOutput())) / ItemStackUtils.getSize(storage.getPrimaryOutput()));
+            }
+        }
+
+        return craftingCount;
+    }
+
+    @Nullable
+    protected List<IToken<?>> createRequestsForRecipe(@NotNull final IRequestManager manager, @NotNull final AbstractBuildingWorker building, final ItemStack requestStack, @NotNull final IRecipeStorage storage)
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Remove usage of generic wildcard type. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1452)
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Split this 210 characters long line (which is greater than 180 authorized). [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS00103)